### PR TITLE
Diagram auto-layout with Graphviz

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [x] Feature
- [ ] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?

Currently we have no way to automatically layout a diagram. This is a feature we probably want if we want to allow textual input and diagram generation.

Issue Number: N/A

### What is the new behavior?

I've been playing around with Graphviz as layout engine.

Nested elements: https://stackoverflow.com/questions/2012036/graphviz-how-to-connect-subgraphs

### Other information

I've been looking at PyNSource, but the license (GPL3) does not allow us to include parts of his code.

This feature could ease implementation of #615 and #1363.

Packaging of Graphviz is handled by #1797.

### CODE REVIEW

**Comment by amolenaar** [comment]
For bundling:

* https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/86
* https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/357

Install Graphviz via Brew/Choco/Apt. Flatpak should build from source.

**Comment by amolenaar** [comment]
It provides acceptable output (and does not crash).
There are 2 things I think can improve the diagrams:

1. Use digraph and `->` to direct head and tail.
2. provide the existing `pos` to elements to guide placement.

**Comment by amolenaar** [comment]
@danyeaw Still graphviz (`dot[.exe]`) needs to be bundled and made sure to work. But you can give it a try. Since diagrams are generated they look less fancy compared to hand drawn diagrams. On the other hand, this allows us to look at more code-centric workflows. Ergo, it's a nice utility.

File: gaphor/plugins/autolayout/tests/test_pydot.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    connect(gen, gen.head, subclass)
+
+    auto_layout = AutoLayout(event_manager, None)
+    auto_layout.layout(diagram)
+
+    assert gen.head.pos != (0, 0)
+    assert gen.tail.pos != (0, 0)
+
+
+def test_layout_with_association(diagram, create, event_manager):
```
Comment (danyeaw) [inline gaphor/plugins/autolayout/tests/test_pydot.py:35]: Does this test, the comment one, and fork node one,  need an assert?

**Review by danyeaw — COMMENTED** [review]
This looks great. Having automatic layout in Gaphor is a huge contribution!

File: gaphor/plugins/autolayout/tests/test_pydot.py (hunk 1 above)
Comment (amolenaar) [reply gaphor/plugins/autolayout/tests/test_pydot.py:35]: The alternative is that an except would be raised. What I wanted to assert is that it does not crash ;). It's hard to make asserts about the positions the elements are located at after auto-layout.

**Comment by amolenaar** [comment]
Added one more commit: the auto-layout feature will only be exposed if `dot` can be found.

**Comment by amolenaar** [comment]
BTW. Instead of pydot, we can also use [Graphviz' python API](https://graphviz.org/pdf/gv.3python.pdf). That might fix our dependency issue.

One other issue to deal with are the plugins. Those are compiled and installed on a "fixed" location.

**Review by danyeaw — APPROVED** [review]

**Pull request merged**